### ISSUE TITLE
Capture and log stdout

### ISSUE THREAD

**Issue by mainej** [clojure-lsp/lsp4clj#9]
The cause of https://github.com/clojure-lsp/clojure-lsp/issues/1080 was that clojure-lsp was accidentally writing to stdout. This was found in https://github.com/clojure-lsp/clojure-lsp/pull/1083 and fixed in https://github.com/clojure-lsp/clojure-lsp/pull/1087.

@ericdallo suggested the following changes in lsp4clj to prevent similar problems in the future:

1. lsp4clj should wrap all of its calls into clojure-lsp and other servers with `with-out-str`. This is because cloure-lsp uses `System.out` to communicate with clients (and I believe all other servers do too). If the server accidentally prints to stdout while it's processing a request, `with-out-str` will avoid letting that output leak into the server/client communication channel.
2. lsp4clj should log any non-blank output strings.

**Comment by ericdallo** [comment]
`with-out-str` may not fit well for us as it returns the string that is on the *out*, so something like: `(with-out-str {:foo 1})` would return `""` and not the map we want to respond in a request for example.
So maybe to fix 1. and 2. using `binding` `*out*` to a local and just logging would fix both 1. and 2.

**Comment by mainej** [comment]
> using `binding` `*out* `to a local and just logging 

@ericdallo I experimented with that idea and it works.

```clojure
(defmacro logging-out-str
  "Evaluates exprs in a context in which *out* is bound to a fresh StringWriter.
  Logs any output, and returns the result of the exprs."
  [& body]
  `(let [s# (java.io.StringWriter.)]
     (binding [*out* s#]
       (let [result# (do ~@body)
             out# (str s#)]
         (when (seq out#)
           (logger/warn "captured stdout: " out#))
         result#))))
```

One thing to note is that if the body starts any threads (e.g., `future` or `go`) which print sometime later then their output is dropped entirely—it's neither logged nor sent to stdout. This was surprising to me at first, but after thinking about it more makes sense. Those threads continue to use the StringWriter for stdout, but the `logger/warn` has run already, capturing the value of the StringWriter at one moment in time before the threads finished.

This should help a lot, but I'm not certain we'll catch everything. For example, `server/run-server!` spawns all those async tasks. That's outside of lsp4clj, so lsp4clj won't capture any prints to stdout that happen in those tasks. That includes some pretty important (and potentially noisy) stuff like `f.file-management/analyze-changes`.

We can make the lsp4clj helper public so we can use it in clojure-lsp when we think it's necessary. But it'll be hard to be confident we've found everything.
